File: translate-bubble/src/main/java/com/fortysevendeg/translace_bubble/ui/components/CloseView.java
package com.fortysevendeg.translace_bubble.ui.components;

import android.content.Context;
import android.graphics.*;
import android.util.AttributeSet;
import android.view.View;
import com.fortysevendeg.translace_bubble.R;

public class CloseView extends View {

    private Paint paintBackground;

    private int width;

    private int height;

    private int middleWidth;

    private int middleHeight;

    private float[] positionsColors = new float[]{0f, 1f};

    private int[] colors = new int[]{Color.parseColor("#99000000"), Color.TRANSPARENT};

    private int radius;

    private int sizeAcross;

    private Paint paintCircle;

    public CloseView(Context context) {
        super(context);
        init(context);
    }

    public CloseView(Context context, AttributeSet attrs) {
        super(context, attrs);
        init(context);
    }

    public CloseView(Context context, AttributeSet attrs, int defStyleAttr) {
        super(context, attrs, defStyleAttr);
        init(context);
    }

    private void init(Context context) {
        paintBackground = new Paint();

        radius = (int) context.getResources().getDimension(R.dimen.radius_close);
        sizeAcross = radius / 3;
        int stroke = (int) context.getResources().getDimension(R.dimen.stroke_close);

        paintCircle = new Paint();
        paintCircle.setColor(Color.WHITE);
        paintCircle.setStyle(Paint.Style.STROKE);
        paintCircle.setStrokeWidth(stroke);
        paintCircle.setAntiAlias(true);

    }

    private void load() {
        if (width == 0) {
            width = getWidth();
            height = getHeight();
            middleWidth = width / 2;
            middleHeight = height / 2;
            paintBackground.setShader(new LinearGradient(width / 2, height, width / 2, 0, colors, positionsColors, Shader.TileMode.CLAMP));
        }
    }

    @Override
    protected void onDraw(Canvas canvas) {
        super.onDraw(canvas);
        load();
        canvas.drawRect(0, 0, width, height, paintBackground);
        canvas.drawCircle(middleWidth, middleHeight, radius, paintCircle);
        canvas.drawLine(middleWidth - sizeAcross, middleHeight - sizeAcross, middleWidth + sizeAcross, middleHeight + sizeAcross, paintCircle);
        canvas.drawLine(middleWidth + sizeAcross, middleHeight - sizeAcross, middleWidth - sizeAcross, middleHeight + sizeAcross, paintCircle);
    }

    public void show() {
        setVisibility(VISIBLE);
    }

    public void hide() {
        setVisibility(GONE);
    }

    public boolean isVisible() {
        return getVisibility() == VISIBLE;
    }

}

File: src/main/scala/com/fortysevendeg/translatebubble/ui/components/CloseView.scala
/*
 * Copyright (C) 2015 47 Degrees, LLC http://47deg.com hello@47deg.com
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may
 * not use this file except in compliance with the License. You may obtain
 * a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.fortysevendeg.translatebubble.ui.components

import android.content.Context
import android.graphics._
import android.util.AttributeSet
import android.view.View
import com.fortysevendeg.translatebubble.R

class CloseView(context: Context, attrs: AttributeSet, defStyleAttr: Int)
    extends View(context, attrs, defStyleAttr) {

  def this(context: Context) = this(context, null, 0)

  def this(context: Context, attr: AttributeSet) = this(context, attr, 0)

  val stroke: Int = context.getResources.getDimension(R.dimen.stroke_close).toInt

  var width: Int = 0
  var height: Int = 0
  var middleWidth: Int = 0
  var middleHeight: Int = 0

  val paintStroke: Paint = {
    val paint = new Paint
    paint.setColor(Color.WHITE)
    paint.setStyle(Paint.Style.STROKE)
    paint.setStrokeWidth(stroke)
    paint.setAntiAlias(true)
    paint
  }

  val paintFillCircle: Paint = {
    val paint = new Paint
    paint.setColor(Color.BLACK)
    paint.setStyle(Paint.Style.FILL)
    paint.setAntiAlias(true)
    paint.setAlpha(50)
    paint
  }

  private def load() {
    if (width == 0) {
      width = getWidth
      height = getHeight
      middleWidth = width / 2
      middleHeight = height / 2
    }
  }

  protected override def onDraw(canvas: Canvas) {
    super.onDraw(canvas)
    load()
    canvas.drawCircle(middleWidth, middleHeight, (width / 2) - stroke, paintFillCircle)
    canvas.drawCircle(middleWidth, middleHeight, (width / 2) - stroke, paintStroke)
    val sizeAcross = width / 6
    canvas.drawLine(middleWidth - sizeAcross, middleHeight - sizeAcross, middleWidth + sizeAcross, middleHeight + sizeAcross, paintStroke)
    canvas.drawLine(middleWidth + sizeAcross, middleHeight - sizeAcross, middleWidth - sizeAcross, middleHeight + sizeAcross, paintStroke)
  }

}
